Combine the vm_run_with_inline_stack and  vm_run_with_alloca

jiangzidong opened the pull request:
Since both _inline_stack and _alloca **use the space of stack**, why don't we combine those 2 branch together? 

Before, when call_stack_size<INLINE_STACK_SIZE, we allocate an array with fixed length INLINE_STACK_SIZE, it also leads the waste of stack. 

But when I run run-perf-test, I cant see the changes of RSS.  @LaszloLango Could you help me measure the performance in your internal tool? Especially the "stack" size

jiangzidong commented:
I did a simple example
```
function f1(){
    f2();
}
function f2(){
    f3();
}
function f3(){
    f4();
}
function f4(){
}
f1();
```

the stack length between first `vm_run` (run the whole script) and fifth `vm_run` (run function f4)
original: 4032 byte
patch: 3840 byte

zherczeg commented:
The motivation for the original approach that allocating a fixed stack size is "free", while allocating dynamic stack is costly. Do you know how dynamic allocation affects other compiler optimizations? How is it exactly allocated? If it turns out to be cheap we can drop the fixed allocation.

jiangzidong commented:
@zherczeg I just objdump the asm code of dymanic array in x86 platfom(gcc, ubuntu). Yes, it will add  more instructions to calc the location of the array, comparing to the fixed array, though both arrays are stored in the stack. 

I think it may be a trade-off between stack-size and perf, because fixed-length array may waste stack space for small `call_stack_size`. The stack resource is limited in many embedded platforms. 

It is just a guess, I am still looking for a tool to measure the runtime maximal stack size of process (jerry). Do you have such tools?

zherczeg commented:
We do it in a dirty way. When we want to measure the stack size, we apply a patch which: creates a large static array, fill it with 0xfe bytes, and set the stack pointer to the top of this array using assembly code. Then it runs jerry, and at the end we check the 0xfe bytes from the bottom. If you wish we can share this patch, but as I mentioned it is a dirty way.

The Heimdall tool of Freya could also be used to measure the stack size (with 64 byte precision):
https://github.com/szeged/Freya

The highest written address range is the stack.

LaszloLango commented:
RPi2 results:

Benchmark | Stack (bytes) | RSS (bytes) | Perf (sec) |
----: | ----: | ----: | ----: | 
3d-cube.js | 17472 -> 16648 : +4.716% | 57344 -> 53248 : +7.143% | 0.938 -> 0.933 : +0.599% | 
3d-raytrace.js | 3348 -> 3348 : 0.000% | 217088 -> 143360 : +33.962% | 1.118 -> 1.107 : +1.043% | 
access-binary-trees.js | 4332 -> 4132 : +4.617% | 53248 -> 49152 : +7.692% | 0.578 -> 0.573 : +0.780% | 
access-fannkuch.js | 2052 -> 2032 : +0.975% | 24576 -> 20480 : +16.667% | 2.386 -> 2.357 : +1.229% | 
access-nbody.js | 2216 -> 2144 : +3.249% | 28672 -> 24576 : +14.286% | 1.082 -> 1.073 : +0.868% | 
bitops-3bit-bits-in-byte.js | 1992 -> 1992 : 0.000% | 20480 -> 12288 : +40.000% | 0.581 -> 0.580 : +0.270% | 
bitops-bits-in-byte.js | 1992 -> 1992 : 0.000% | 20480 -> 12288 : +40.000% | 0.861 -> 0.862 : -0.096% | 
bitops-bitwise-and.js | 1768 -> 1768 : 0.000% | 20480 -> 12288 : +40.000% | 1.095 -> 1.070 : +2.251% | 
bitops-nsieve-bits.js | 1992 -> 1992 : 0.000% | 155648 -> 151552 : +2.632% | 1.774 -> 1.751 : +1.278% | 
controlflow-recursive.js | 76536 -> 72440 : +5.352% | 94208 -> 81920 : +13.043% | 0.408 -> 0.388 : +4.955% | 
crypto-aes.js | 3376 -> 3352 : +0.711% | 90112 -> 90112 : 0.000% | 1.095 -> 1.086 : +0.863% | 
crypto-md5.js | 2604 -> 2524 : +3.072% | 167936 -> 159744 : +4.878% | 0.735 -> 0.729 : +0.763% | 
crypto-sha1.js | 2072 -> 2072 : 0.000% | 114688 -> 106496 : +7.143% | 0.706 -> 0.697 : +1.251% | 
date-format-tofte.js | 2784 -> 2696 : +3.161% | 40960 -> 36864 : +10.000% | 0.849 -> 0.849 : +0.017% | 
date-format-xparb.js | 3096 -> 3064 : +1.034% | 45056 -> 36864 : +18.182% | 0.453 -> 0.446 : +1.577% | 
math-cordic.js | 1992 -> 1992 : 0.000% | 24576 -> 16384 : +33.333% | 1.339 -> 1.328 : +0.824% | 
math-partial-sums.js | 1944 -> 1944 : 0.000% | 20480 -> 12288 : +40.000% | 0.759 -> 0.727 : +4.279% | 
math-spectral-norm.js | 2268 -> 2180 : +3.880% | 24576 -> 20480 : +16.667% | 0.601 -> 0.597 : +0.794% | 
string-base64.js | 2000 -> 2000 : 0.000% | 110592 -> 131072 : -18.519% | 1.928 -> 1.914 : +0.720% | 
string-fasta.js | 1976 -> 1976 : 0.000% | 32768 -> 28672 : +12.500% | 1.412 -> 1.389 : +1.617% | 
Geometric mean: | +1.557% | +18.523% | +1.302% | 

Binary sizes (bytes)
971dcae669:154136
c3f5b89a28:156692

jiangzidong commented:
@zherczeg I tried the freya, but I don't know which data means stack size (I think it is measuring jerry-heap usage). Do I need some extra configuration?

1, build jerry with --valgrid-freya=on
2,  `./inst/bin/valgrind --tool=freya ~/jerryscript/build/bin/jerry ~/jerryscript/debug.js`
3, get result:
```
Trace:

[0] Peak: 724b Allocs: 40 Reallocs: 0 Total: 1.4Kb (1436b)
Group: total
  [1] (/home/jzd/jerryscript/build/bin/jerry)
    [2] Peak: 560b Allocs: 15 Reallocs: 0 Total: 1.0Kb (1112b)
    (/home/jzd/jerryscript/build/bin/jerry)
      [3] Peak: 240b Allocs: 5 Reallocs: 0 Total: 600b
      (/home/jzd/jerryscript/build/bin/jerry)
        [4] Peak: 240b Allocs: 4 Reallocs: 0 Total: 480b
        (/home/jzd/jerryscript/build/bin/jerry)
          [5] Peak: 120b Allocs: 3 Reallocs: 0 Total: 360b
          (/home/jzd/jerryscript/build/bin/jerry)
          [5] Peak: 120b Allocs: 1 Reallocs: 0 Total: 120b
          (/home/jzd/jerryscript/build/bin/jerry)
...............
```

zherczeg commented:
please use the heimdall tool:
--tool=heimdall --page_size=64

zherczeg commented:
Anyway the results seems convincing, so I think this patch could go in.

zherczeg commented:
LGTM

jiangzidong commented:
@zherczeg Thanks, but I don't understand why perf is also improved.... 

I tried the heimdall, and could you tell me the meanings of 4 rows? 

3d-cube.js
```
Page size: 64 bytes
  Modified area: 0x63a000 - 0x63f040 size: 20544 bytes [20.0 Kb]
  Modified area: 0x6bb000 - 0x6bb200 size: 512 bytes [0.5 Kb]
  Modified area: 0xffeff9100 - 0xffeffff00 size: 28160 bytes [27.5 Kb]
  Modified area: 0xfff0002c0 - 0xfff000340 size: 128 bytes [0.1 Kb]
Total modified memory area: 49344 bytes [48.1 Kb]
```

zherczeg commented:
These are the address spaces written by your application. Heimdall only tracks memory writes.

Modified area: 0xffeff9100 - 0xffeffff00 size: 28160 bytes [27.5 Kb] - high address usually represents the stack so 27.5K is the stack consumption (stack is top-down on x86 and ARM).

Modified area: 0x63a000 - 0x63f040 size: 20544 bytes [20.0 Kb] - this is likely the jerry heap memory consumption:

Other spaces might be some other globals or ld used it for something. You need to check the map file generated by the compiler to know exactly the static areas.

Total modified memory area: 49344 bytes [48.1 Kb] - this is the total written memory (heimdall only track memory writes). So your application requires this amount of memory to run although increased GC activity could reduce this.

jiangzidong commented:
@zherczeg I see, thanks very much for your explanation! :)
